docs/design/metainfo-rpc.md

```diff
@@ -26,87 +26,211 @@ The following is a proto file for just the rpc portion to create segments. The r
+message BucketCreateResponse {}
 
-    // Start Segment
-    int64 segment_index = 4;
-    int64 part_number = 5;
+message BucketGetRequest {
+    bytes name = 1;
+}
 
-    // the following copied from an earlier version of these requests
-    pointerdb.RedundancyScheme redundancy = 6;
-    int64 max_encrypted_segment_size = 7;
-    google.protobuf.Timestamp expiration = 8;
+message BucketGetResponse {
+    Bucket bucket = 1;
 }
 
-message CommitSegmentAndStartSegmentRequest {
-    // Commit Segment
-    bytes segment_upload_id = 1;
-    pointerdb.Pointer pointer = 2;
-    repeated orders.OrderLimit2 original_limits = 3;
+message BucketDeleteRequest {
+    bytes name = 1;
+}
+message BucketDeleteResponse {}
 
-    // Start Segment
-    bytes stream_upload_id = 4;
-    int64 segment_index = 5;
-    int64 part_number = 6;
+message BucketListRequest {
+    bytes     cursor    = 1;
+    Direction direction = 2;
+    int32     limit     = 3;
+}
 
-    // The following copied from an earlier version of these requests
-    pointerdb.RedundancyScheme redundancy = 7;
-    int64 max_encrypted_segment_size = 8;
-    google.protobuf.Timestamp expiration = 9;
+message BucketListResponse {
+    repeated Bucket items = 1;
+    bool            more  = 2;
 }
 
-message CommitSegmentAndCommitStreamRequest {
-    // Commit Segment
-    bytes segment_upload_id = 1;
-    pointerdb.Pointer pointer = 2;
-    repeated orders.OrderLimit2 original_limits = 3;
+message Object {
+    enum Status {
+        UPLOADING;
+        COMMITTING;
+        COMMITTED;
+        DELETING;
+    }
 
-    // Commit Stream
-    bytes stream_upload_id = 4;
+    bytes  bucket;
+    bytes  encrypted_path;
+    int32  version;
+    Status status;
+
+    bytes  stream_id;
+
+    google.protobuf.Timestamp created_at;
+    google.protobuf.Timestamp status_at;
+    google.protobuf.Timestamp expires_at;
+
+    bytes  encrypted_metadata_nonce;
+    bytes  encrypted_metadata;
+
+    int64                fixed_segment_size;
+    RedundancyScheme     redundancy_scheme;
+    EncryptionParameters encryption_parameters;
+
+    int64 total_size;
+    int64 inline_size;
+    int64 remote_size;
 }
 
-message StartStreamAndStartSegmentResponse {
-    // Start Stream
-    bytes stream_upload_id = 1;
+message ObjectCreateRequest {
+    bytes  bucket = 1;
+    bytes  encrypted_path = 2;
+    int32  version = 3;
+}
+
+message ObjectCreateResponse {
+    bytes  bucket;
+    bytes  encrypted_path;
+    int32  version;
+
+    bytes  stream_id;
+
+    google.protobuf.Timestamp expires_at;
+
+    bytes  encrypted_metadata_nonce;
+    bytes  encrypted_metadata;
+
+    RedundancyScheme     redundancy_scheme;
+    EncryptionParameters encryption_parameters;
+}
 
-    // Start Segment
-    bytes segment_upload_id = 2;
-    repeated AddressedOrderLimit addressed_limits = 3;
-    bytes root_piece_id = 4 [(gogoproto.customtype) = "PieceID", (gogoproto.nullable) = false];
+message ObjectCommitRequest {
```

Comment:
I would prefer if this took the returned stream_id so that the server has the ability to use it like a browser cookie. It already has to depend on the create part to use the absolute version number reported from the response, and it seems like an easy bug to accidentally have it use `-1` for both that can be avoided entirely.

Reply:
wait why would the commit with version = -1 fail?

Reply:
Because perhaps a new version was started during your upload, and so -1 means something different than when you started uploading.

Reply:
How this kind of business logic errors are reported/responded in gRPC?

Reply:
I don't think this requires an additional index in the same way that the segment_upload_id wouldn't. The server can store the information as an opaque byte sequence that the client would send here, anyway.